dropshot/src/handler.rs

```diff
@@ -1065,57 +1078,130 @@ where
+    ///
+    /// Recommended for smaller request bodies. Constructing a single `Bytes`
+    /// chunk from larger request bodies might cause excessive copying and
+    /// reallocations.
+    ///
+    /// # Errors
+    ///
+    /// Errors if the request body is too large, or if the
+    pub async fn into_bytes(self) -> Result<Bytes, HttpError> {
+        let mut stream = self.into_stream();
+        let mut bytes = BytesMut::new();
+
+        while let Some(data) = stream.next().await {
+            bytes.put(data?);
+        }
+
+        Ok(bytes.freeze())
     }
 
-    /**
-     * Convenience wrapper to convert the body to a UTF-8 string slice,
-     * returning a 400-level error if the body is not valid UTF-8.
-     */
-    pub fn as_str(&self) -> Result<&str, HttpError> {
-        std::str::from_utf8(self.as_bytes()).map_err(|e| {
+    /// Reads the body into a `String`.
+    ///
+    /// # Errors
+    ///
+    /// In addition to the usual errors, if the stream is not valid UTF-8, this
+    /// returns a "Bad Request" error.
+    pub async fn into_string(self) -> Result<String, HttpError> {
+        let v = Vec::from(self.into_bytes().await?);
+        String::from_utf8(v).map_err(|e| {
             HttpError::for_bad_request(
                 None,
                 format!("failed to parse body as UTF-8 string: {}", e),
             )
         })
     }
+
+    /// Reads the body into a [`BufList`].
+    ///
+    /// A `BufList` is a list of [`Bytes`] chunks that implements the
+    /// [`Buf`](bytes::Buf) trait. A `BufList` chunks can be operated on
+    /// as a unit.
+    ///
+    /// Recommended for larger request bodies.
+    pub async fn into_buf_list(self) -> Result<BufList, HttpError> {
+        let max_bytes = self.max_bytes;
+        self.into_buf_list_with_cap(max_bytes).await
+    }
+
+    /// Reads the body into a [`BufList`] with a custom limit for the maximum
+    /// size of bytes.
+    ///
+    /// This method is similar to [`into_buf_list`](Self::into_buf_list), except
+    /// a custom limit is specified. If this method is called, the default
+    /// [`request_body_max_bytes`](ServerConfig::request_body_max_bytes) limit
+    /// is ignored.
+    pub async fn into_buf_list_with_cap(
+        self,
+        max_bytes: usize,
+    ) -> Result<BufList, HttpError> {
+        let mut stream =
+            self.into_stream().into_uncapped().into_capped(max_bytes);
+        let mut buf_list = BufList::new();
+
+        while let Some(data) = stream.next().await {
+            buf_list.push_chunk(data?);
+        }
+
+        Ok(buf_list)
+    }
+
+    /// Converts `self` into a [`Stream`] of `Result<Bytes, HttpError>` chunks.
+    pub fn into_stream(self) -> CappedBodyStream {
+        let max_bytes = self.max_bytes;
+
+        let stream = stream! {
+            let mut request = self.request.lock().await;
+            let body = request.body_mut();
+
+            'outer: {
```

Comment:
I know I said I'd reserve my feedback until you go through @davepacheco's comments, but free advice on this: it might be less work to defer the use of relatively-new structures rather than potentially impacting some dropshot consumer at Oxide that's on an older rust version. My personal threshold is about 4-6 months in terms of my willingness to just hope that folks are up-to-date.